runtime/sema/check_assignment.go

```diff
@@ -107,6 +110,76 @@ func (checker *Checker) checkAssignment(
 	return
 }
 
+func (checker *Checker) enforcePureAssignment(assignment ast.Statement, target ast.Expression) {
+	if !checker.CurrentPurityScope().EnforcePurity {
+		return
+	}
+
+	var variable *Variable
+
+	switch targetExp := target.(type) {
+	case *ast.IdentifierExpression:
+		variable = checker.valueActivations.Find(targetExp.Identifier.Identifier)
+	case *ast.IndexExpression:
+		if indexIdentifier, ok := targetExp.TargetExpression.(*ast.IdentifierExpression); ok {
+			variable = checker.valueActivations.Find(indexIdentifier.Identifier.Identifier)
+		}
+	case *ast.MemberExpression:
+		if indexIdentifier, ok := targetExp.Expression.(*ast.IdentifierExpression); ok {
+			variable = checker.valueActivations.Find(indexIdentifier.Identifier.Identifier)
+		}
+	}
+
+	// if the target is not a variable (e.g. a nested index expression x[0][0], or a nested
+	// member expression x.y.z), the analysis cannot know for sure where the value being
+	// assigned to originated, so we must default to impure. Consider:
+	// -----------------------------------------------------------------------------
+	// let a: &[Int] = &[0]
+	// pure fun foo(_ x: Int) {
+	//     let b: &[Int] = [0]
+	//     let c = [a, b]
+	//     c[x][0] = 4 // we cannot know statically whether a or b receives the write here
+	// }
+	if variable == nil {
+		checker.ObserveImpureOperation(assignment)
+		return
+	}
+
+	// `self` technically exists in param scope, but should still not be writeable
```

Comment:
Does this imply that initializers assigning to fields of `self` are considered pure?

Reply:
Yes, for the following reason: whenever we call a constructor from inside a pure scope, the value being constructed is local to that scope, so it is safe to create a new value from within a pure scope. This means that functions that just construct new values are technically pure (in the same way that they are in a functional programming sense), as long as they don't modify anything else while constructing those values.

Reply:
Ah, right!